Remove WaitRunning

WeiZhang555 opened the pull request:
Currently SetRunning and SetStopped share same wait channel, this can
lead to a wait operation for running state getting wrong state stopped.

This commit will create two separate channels for Running&Stopped states.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

LK4D4 commented:
btw I'm not convinced on a sole usage of `WaitRunning` in `daemon/update.go`. I'd prefer to use just state polling there and remove complexity from `container/state.go`.

WeiZhang555 commented:
@LK4D4 Yep, good idea. It's better to remove waitRunning completely rather than separate wait channel. :+1:

WeiZhang555 committed:
Remove WaitRunning

Remove function `WaitRunning` because it's actually not necessary, also
remove wait channel for state "running" to avoid mixed use of the state
wait channel.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
I've removed function "WaitRunning" to remove the complexity, it looks better. @LK4D4

WeiZhang555 commented:
And I removed a lot from your unit test codes, as WaitRunning is removed by me, they are not necessary too.

LK4D4 commented:
@WeiZhang555 I think some tests hanging. Pls try locally tests about daemon update.

WeiZhang555 commented:
All green now. Failed test case is not related. :smile: 
@LK4D4 PTAL

WeiZhang555 commented:
ping @LK4D4

LK4D4 commented on daemon/update.go:
```diff
@@ -74,11 +73,6 @@ func (daemon *Daemon) update(name string, hostConfig *container.HostConfig) erro
 	container.UpdateMonitor(hostConfig.RestartPolicy)
 
-	// if container is restarting, wait 5 seconds until it's running
-	if container.IsRestarting() {
```
Probably we should just poll here for some time. Just without `WaitRunning`.

WeiZhang555 replied:
Actually this piece of code was first introduced by me, I was hoping to let container reach a stable state(running/stopped) before trying to update the container's resources. After some consideration, I think even we add polling (say 5 seconds), it can't guarantee the container can escape from the restarting state, maybe it can decrease the risk of race but can't remove the risk all. 

So I think it's safe to remove this, and save the blocking time then user can get a quicker feedback.

What do you think?

LK4D4 replied:
Ok, sounds reasonable.

LK4D4 commented:
LGTM

thaJeztah commented:
ping @tonistiigi @cpuguy83 PTAL

WeiZhang555 commented:
ping @tonistiigi @cpuguy83 @calavera 
Could you have a look at this?

tonistiigi commented:
LGTM
